---
title: "costal_ots_NRT_analysis"
author: "Dani Blumstein"
date: "2024-11-14"
output: html_document
---

```{r}
library(adegenet)
library(poppr)
library(tidyverse)
library(readxl)
library(adegenet)
library(dartR)
library(RColorBrewer)
library(igraph)
library(hierfstat)
library(reshape2)
library(ggplot2)
library(scales)
library(devtools)
library(splitstackshape)
library(cowplot)
library(gridGraphics)
library(ggrepel)
library(magrittr)
library(paletteer)

if (!require("remotes")) install.packages("remotes")
remotes::install_github("hhwagner1/LandGenCourse")
library(LandGenCourse)

if(!requireNamespace("LandGenCourseData", quietly = TRUE))
            devtools::install_github("hhwagner1/LandGenCourseData")

if(!requireNamespace("fields", quietly = TRUE)) install.packages("fields", repos='http://cran.us.r-project.org')
  
if(!requireNamespace("LEA", quietly = TRUE)) {  
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  BiocManager::install("LEA")
}

library(LEA)
library(scatterpie)
library(ape)
library(phangorn)
library(ggnewscale)
```

read in the data
```{r}
meta<-read.csv("Ots_coastal_2024_meta.csv", header = TRUE)
genos_all <- read.csv("Ots_coastal_2024_loci_data.csv", header = TRUE)

##read marker information for Ots GT-seq panel
ots_panel_info<-read_excel("loci_details.xlsx")
```

```{r}
#COLORS!! because colors are fun
hcl.colors(18, "Zissou1", rev = FALSE)

group.colors <- c(CHER = "#3B99B1", COOR = "#36A4AB", COQR ="#45ADA3", ELKR = "#5FB49B", KLAR = "#7CBA96", NESR ="#97BE95", NUMP = "#ADC491", ROGR= "#C4C88B", SILR= "#EACB2B", SIUR= "#EABC22", SIXR= "#E9AE1D", SUMP= "#E89F16", TILR= "#E78F0A", TRAR= "#E77E00", TRIR= "#EA6900", UMPR= "#EE4E00", WILR= "#F5191C", YAQR= "#F5191C")

group.colors2 <- c("Chetco River" = "#3B99B1", "Coos River" = "#36A4AB", "Coquille River" ="#45ADA3", "Elk River" = "#5FB49B", "Klamath River" = "#7CBA96", "Nestucca River" ="#97BE95", "Little Nestucca River" = "#97BE95", "Rogue River"= "#C4C88B", "Siletz River"= "#EACB2B", "Siuslaw River"= "#EABC22", "Sixes River"= "#E9AE1D", "South Umpqua River"= "#E89F16", "Tillamook River"= "#E78F0A", "Trask River"= "#E77E00", "Trinity River"= "#EA6900", "North Umpqua River"= "#ADC491", "Umpqua River" = "#EE4E00","Wilson River"= "#F5191C", "Yaquina River"= "#F5191C")
                  
```


```{r}
##put dataset and meta in same order##
meta<-meta[order(match(meta$sample, genos_all$sample)), ]
identical(genos_all$sample, meta$sample)

##subset panel info to those markers in the SFGL panel
ots_SFGL_panel_info<-ots_panel_info[ots_panel_info$marker %in% colnames(genos_all), ]
```

look at 218 markers identified as Neutral
```{r}
nm<-which(ots_panel_info$`Presumed Type` == "Neutral")
neutral_marker_list<-ots_panel_info[nm, ]
neutral_matrix<-as.matrix(genos_all[, colnames(genos_all) %in% neutral_marker_list$marker])
rownames(neutral_matrix)<-genos_all$sample

##remove "dots" from column names##
colnames(neutral_matrix) <- gsub("\\.", "_", colnames(neutral_matrix))

neutral_genind <- df2genind(neutral_matrix, sep ="", ploidy=2,NA.char = "00")
neutral_genind
```

```{r}
pop(neutral_genind) <- meta$Population

# Conversion to genclone
genclone_nutral_markers <- as.genclone(neutral_genind)
#strata(genclone_nutral_markers) <- meta
genclone_nutral_markers$pop<-as.factor(meta$Population)
#splitStrata(genclone_nutral_markers) <- ~origin/run

# Conversion to genlight 
neutral_genlight <- gi2gl(neutral_genind, parallel = FALSE, verbose = NULL)
ploidy(neutral_genlight) <- 2
pop(neutral_genlight) <- meta$Population

# Conversion to geno
gl2geno(neutral_genlight, outfile = "neutral", outpath = getwd(), verbose = NULL)

# Conversion to genpop
neutral_genpop <- genind2genpop(neutral_genind)


#better way to make allllll the objects
#require(strataG)
#snowcrab <- genomic_converter(
#                   data = "populations.snps.vcf", strata = "snowcrab.strata.tsv",
#                   output = c("genlight", "genepop", "gtypes"))
```

summary stats by pop
```{r}
#alleles per locus
table(neutral_genind$loc.fac)

#samples per site
summary(neutral_genind$pop)

#private alleles per site across all loci
#private_alleles(neutral_genind) %>% apply(MARGIN = 1, FUN = sum)

#mean allelic richness per site across all loci
allelic.richness(genind2hierfstat(neutral_genind))$Ar %>%
  apply(MARGIN = 2, FUN = mean) %>% 
  round(digits = 3)

#heterozygosity per site
basic_neutral = basic.stats(neutral_genind, diploid = TRUE)

#Mean observed heterozygosity per site
Ho_neutral = apply(basic_neutral$Ho, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 2)
Ho_neutral

# Mean expected heterozygosity per site
He_neutral = apply(basic_neutral$Hs, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 2)
He_neutral

Het_neutral_df = data.frame(Site = names(Ho_neutral), Ho = Ho_neutral, He = He_neutral) %>%
  melt(id.vars = "Site")

# Custom theme for ggplot2
custom_theme = theme(
  axis.text.x = element_text(size = 10, angle = 90, vjust = 0.5, face = "bold"),
  axis.text.y = element_text(size = 10),
  axis.title.y = element_text(size = 12),
  axis.title.x = element_blank(),
  axis.line.y = element_line(size = 0.5),
  legend.title = element_blank(),
  legend.text = element_text(size = 12),
  panel.grid = element_blank(),
  panel.background = element_blank(),
  plot.title = element_text(hjust = 0.5, size = 15, face="bold")
  )

# Italic label
hetlab.o = expression(italic("H")[o])
hetlab.e = expression(italic("H")[e])

ggplot(data = Het_neutral_df, aes(x = Site, y = value, fill = variable))+
  geom_bar(stat = "identity", position = position_dodge(width = 0.6), colour = "black")+
  scale_y_continuous(expand = c(0,0), limits = c(0,0.50))+
  scale_fill_manual(values = c("royalblue", "#bdbdbd"), labels = c(hetlab.o, hetlab.e))+
  ylab("Heterozygosity")+
  ggtitle("neutral loci")+
  custom_theme

#mean FIS per site (as oppose to FIS per individual)
apply(basic_neutral$Fis, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 3)

#pairwise FST (Weir & Cockerham 1984)
neutral_fst = genet.dist(neutral_genind, method = "WC84") %>% round(digits = 3)
neutral_fst

lab_order = c("ROGR","NESR","NUMP","SILR","TILR","TRAR","SUMP","COOR","COQR","SIUR","SIXR","UMPR","WILR","YAQR","TRIR","KLAR","CHER")

# Change order of rows and cols
fst.mat = as.matrix(neutral_fst)
fst.mat1 = fst.mat[lab_order, ]
fst.mat2 = fst.mat1[, lab_order]

# Create a data.frame
ind = which(upper.tri(fst.mat2), arr.ind = TRUE)
fst.df = data.frame(Site1 = dimnames(fst.mat2)[[2]][ind[,2]],
                    Site2 = dimnames(fst.mat2)[[1]][ind[,1]],
                    Fst = fst.mat2[ ind ])

# Keep the order of the levels in the data.frame for plotting 
fst.df$Site1 = factor(fst.df$Site1, levels = unique(fst.df$Site1))
fst.df$Site2 = factor(fst.df$Site2, levels = unique(fst.df$Site2))

# Convert minus values to zero
fst.df$Fst[fst.df$Fst < 0] = 0

# Print data.frame summary
fst.df %>% str

# Fst italic label
fst.label = expression(italic("F")[ST])

# Extract middle Fst value for gradient argument
mid = max(fst.df$Fst) / 2

# Plot heatmap
ggplot(data = fst.df, aes(x = Site1, y = Site2, fill = Fst))+
  geom_tile(colour = "black")+
  geom_text(aes(label = Fst), color="black", size = 3)+
  scale_fill_gradient2(low = "blue", mid = "pink", high = "red", midpoint = mid, name = fst.label, limits = c(0, max(fst.df$Fst)), breaks = c(0, 0.05, 0.10, 0.15))+
  scale_x_discrete(expand = c(0,0))+
  scale_y_discrete(expand = c(0,0), position = "right")+
  theme(axis.text = element_text(colour = "black", size = 10, face = "bold"),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.position = "right",
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 10)
        )
```


individual level PCA (missing data filled in with mean genotype)
```{r}
# Replace missing data with the mean allele frequencies
x = tab(neutral_genind, NA.method = "mean")

# Perform PCA
pca1 = dudi.pca(x, scannf = FALSE, scale = FALSE, nf = 3)

# Analyse how much percent of genetic variance is explained by each axis
percent = pca1$eig/sum(pca1$eig)*100
barplot(percent, ylab = "Genetic variance explained by eigenvectors (%)", ylim = c(0,12),
        names.arg = round(percent, 1))

# Create a data.frame containing individual coordinates
ind_coords = as.data.frame(pca1$li)

# Rename columns of dataframe
colnames(ind_coords) = c("Axis1","Axis2","Axis3")

# Add a column containing individuals
ind_coords$Ind = indNames(neutral_genind)

# Add a column with the site IDs
ind_coords$Site = neutral_genind$pop

# Calculate centroid (average) position for each population
centroid = aggregate(cbind(Axis1, Axis2, Axis3) ~ Site, data = ind_coords, FUN = mean)

# Add centroid coordinates to ind_coords dataframe
ind_coords = left_join(ind_coords, centroid, by = "Site", suffix = c("",".cen"))

# Define colour palette
#cols = brewer.pal(nPop(neutral_genind), "Set1")

# Custom x and y labels
xlab = paste("Axis 1 (", format(round(percent[1], 1), nsmall=1)," %)", sep="")
ylab = paste("Axis 2 (", format(round(percent[2], 1), nsmall=1)," %)", sep="")

# Custom theme for ggplot2
ggtheme = theme(axis.text.y = element_text(colour="black", size=12),
                axis.text.x = element_text(colour="black", size=12),
                axis.title = element_text(colour="black", size=12),
                panel.border = element_rect(colour="black", fill=NA, size=1),
                panel.background = element_blank(),
                plot.title = element_text(hjust=0.5, size=15) 
)

# Scatter plot axis 1 vs. 2
ggplot(data = ind_coords, aes(x = Axis1, y = Axis2))+
  geom_hline(yintercept = 0)+
  geom_vline(xintercept = 0)+
  # spider segments
  geom_segment(aes(xend = Axis1.cen, yend = Axis2.cen, color = Site), show.legend = FALSE)+
  # points
  geom_point(aes(fill = Site), shape = 21, size = 3, show.legend = FALSE)+
  # centroids
  geom_label(data = centroid, aes(label = Site, fill = Site), size = 4, show.legend = FALSE)+
  # colouring
  scale_color_manual(values=group.colors)+
  scale_fill_manual(values=group.colors)+
  # custom labels
  labs(x = xlab, y = ylab)+
  ggtitle("neutral PCA")+
  # custom theme
  ggtheme
```

population level PCA
```{r}
# Replace missing data with the mean allele frequencies
x = tab(neutral_genpop)

# Perform PCA
pca2 = dudi.pca(x, scannf = FALSE, scale = FALSE, nf = 3)

# Analyse how much percent of genetic variance is explained by each axis
percent = pca2$eig/sum(pca2$eig)*100
barplot(percent, ylab = "Genetic variance explained by eigenvectors (%)", ylim = c(0,12),
        names.arg = round(percent, 1))

pop_pcs <- pca2$li

#create column of sample IDs from rownames##
pop_pcs %<>%
  rownames_to_column(var = "pop") 

ggplot(pop_pcs)+
  geom_point(aes(Axis1, Axis2, color = pop), alpha = 0.7, size = 5.0)+
  geom_text_repel(aes(x=Axis1, y=Axis2, label = pop), nudge_x = 0.015, nudge_y = 0.015, size = 4, alpha = 0.8, force = 6, max.overlaps = 15, min.segment.length = 0.8, segment.colour = "darkgrey", segment.size = 1) +
  xlab("PC1 97.3%")+
  ylab("PC2 16.3%")+
  scale_color_manual(values=group.colors)
  
```


```{r}
neutral_df <- as.data.frame(neutral_matrix)

delim_gtypes <- neutral_df[, ] %>%
  mutate_all(funs(case_when(. == "AA" ~ "A/A",
                            . == "AT" ~ "A/T", 
                            . == "AC" ~ "A/C", 
                            . == "AG" ~ "A/G", 
                            . == "TT" ~ "T/T", 
                            . == "TA" ~ "T/A", 
                            . == "TC" ~ "T/C", 
                            . == "TG" ~ "T/G",
                            . == "CC" ~ "C/C", 
                            . == "CA" ~ "C/A", 
                            . == "CT" ~ "C/T", 
                            . == "CG" ~ "C/G", 
                            . == "GG" ~ "G/G", 
                            . == "GA" ~ "G/A", 
                            . == "GT" ~ "G/T", 
                            . == "GC" ~ "G/C", 
                            . == "--" ~ "-/-", 
                            . == "A-" ~ "A/-", 
                            . == "T-" ~ "T/-", 
                            . == "C-" ~ "C/-", 
                            . == "G-" ~ "G/-", 
                            . == "-A" ~ "-/A",
                            . == "-T" ~ "-/T", 
                            . == "-C" ~ "-/C", 
                            . == "-G" ~ "-/G", 
                            . == "00" ~ "0/0", 
                            . == "0" ~ "0/0")))


vrtnames<-c(colnames(delim_gtypes))

##split genotypes into 2 alleles
data<-as.data.frame(cSplit(delim_gtypes, splitCols = vrtnames, sep = "/", type.convert = FALSE))

coancestrylbls<-data.frame(c(genos_all$sample))
colnames(coancestrylbls)<-c("Ind")

rownames(data)<-as.character(coancestrylbls$Ind)

##replace nucleotides with numeric values
##A = 1; T = 2; C = 3; G = 4; - = 9 ###
data %<>%
  mutate_all(funs(case_when(. == "A" ~ 1,
                            . == "T" ~ 2,
                            . == "C" ~ 3,
                            . == "G" ~ 4, 
                            . == "-" ~ 9,
                            . == "0" ~ 0,)))


pops_order = c("TRAR", "WILR", "TILR", "NESR", "SILR", "YAQR","SIUR", "UMPR", "NUMP", "SUMP","COOR","COQR", "SIXR","ELKR","ROGR","KLAR","CHER", "TRIR")
sample_sites <- meta$Population
sorted_sites <- unlist(lapply(pops_order, function(x) sample_sites[sample_sites == x]))


N <- unlist(lapply(pops_order,function(x){length(which(sample_sites==x))}))
names(N) <- pops_order

par(mar=c(4,4,2,0.5))
pcaS <- prcomp(data,center=T)
plot(pcaS$sdev^2 / sum(pcaS$sdev^2), xlab="PC",
     ylab="Fraction Variation Explained", main="Scree plot")

perc <- round(100*(pcaS$sdev^2 / sum(pcaS$sdev^2))[1:10],2)
names(perc) <- apply(array(seq(1,10,1)), 1, function(x){paste0("PC", x)})
perc 

sites <- meta[c("Population","lat","lon")]

#data that needs adding to sites
sites$lat[sites$Population == "TRIR"] <- 40.680109
sites$lon[sites$Population == "TRIR"] <- -122.882794

sites$lat[sites$Population == "KLAR"] <- 42.2234
sites$lon[sites$Population == "KLAR"] <- -121.7776

sites$lat[sites$Population == "CHER"] <- 42.2151096
sites$lon[sites$Population == "CHER"] <- -123.8967436

sites$lat[sites$Population == "ELKR"] <- 42.71667
sites$lon[sites$Population == "ELKR"] <- -124.20528

sites$lat[sites$Population == "UMPR"] <- 43.2681707
sites$lon[sites$Population == "UMPR"] <- -123.4459121

sites$lat[sites$Population == "SUMP"] <- 42.9396
sites$lon[sites$Population == "SUMP"] <- -123.26761

sites$lat[sites$Population == "COQR"] <- 43.1237199
sites$lon[sites$Population == "COQR"] <- -124.4301156

sites$lat[sites$Population == "ROGR"] <- 42.434953
sites$lon[sites$Population == "ROGR"] <- -123.170053

sites2 <- sites %>%
  dplyr::group_by(Population) %>%
  dplyr::summarise(
    lat = mean(lat),
    lon = mean(lon)
  )

sites_n <- as.data.frame(table(sites$Population)) 
sites_n %<>% dplyr::rename("Population" = "Var1")
sites_merge <- merge(sites_n,sites2,by="Population") 

sites2$Population <- factor(sites2$Population, levels = pops_order)

#sort it north to south
sites2 <- sites2[order(sites2$Population), ]
sites2 <- subset(sites2, select = -Population)

snmf <- LEA::snmf("neutral.geno", K=1:18, ploidy=2, entropy=T, 
                   alpha=100, project="new")

plot(snmf, col = "blue4", cex = 1.4, pch = 19)

ce <- cross.entropy(snmf, K = 9)

# select the run with the lowest cross-entropy value
best <- which.min(ce)

qmatrix = Q(snmf, K = 9, run = best)
cluster<- apply(qmatrix, 1, which.max)


# plot the ancestry coefficients for the best run and K
par(mar=c(4,4,0.5,0.5))
barplot(t(qmatrix), col=hcl.colors(9, "Fall", rev = FALSE), 
        border=NA, space=0, ylab="Rivers", 
        xlab="Admixture coefficients", horiz = T,)

#Add population labels to the axis:
for (i in 1:length(pops)){
  axis(4, at=median(which(sample_sites==pops[i])), labels=pops[i])}

p2 <- recordPlot()

K=9

#initialize array for ancestry proportions:
qpop <- matrix(NA,nrow=length(pops_order),ncol=K) 

unique(meta$Population)
#intialize array for coordinates:
coord.pop <- matrix(NA,nrow=length(pops_order),ncol=2) 
n = 1

for (i in unique(pops_order))
  {
  qpop[n,] = apply(qmatrix[pops_order == i,], 2, mean)
  coord.pop[n,] = apply(sites2[pops_order == i,], 2, mean)
  n=n+1
}

colnames(coord.pop) <- c("lat", "lon")
pie_data <- cbind(sites_merge,qpop)

p1 <- ggplot() +
  geom_spatraster(data = r2, mapping = aes(fill = USA_elv_msk)) +
  #geom_sf(data = oregon, size = 1.5, color = "grey50", fill = "#777")+
  geom_sf(data = rivers_filtered, aes(color = Name), lwd = 1.2)+
  geom_sf(data = COOR, aes(color = ID), lwd = 1.2)+
  geom_sf(data = SILR, aes(color = ID), lwd = 1.2)+
  geom_sf(data = SIUR, aes(color = ID), lwd = 1.2)+
  geom_sf(data = SIXR, aes(color = ID), lwd = 1.2)+
  scale_color_manual(values=group.colors2) +
  scale_fill_gradient(high = "lightgrey", low = "#555",  na.value='transparent') +
  coord_sf(xlim=c(-125, -121.7), ylim = c(42,46.2991),datum = sf::st_crs(4326))+
  scale_x_continuous(breaks = seq(-125, -121.7, by = 1), name = "Longitude (°W)")+
  scale_y_continuous(name = "Latitude (°N)") +
  
# start a new scale
  new_scale_fill() +

  geom_scatterpie(data = pie_data,mapping = aes(x=lon, y=lat, group=Population, r=log(Freq+1)/40), cols=colnames(pie_data)[5:13], color = "black")+
  scale_fill_manual(values = hcl.colors(9, "Fall", rev = FALSE))+
  theme(legend.justification = "top")+
  annotation_scale(location = "bl", width_hint = 0.5) +
  annotation_north_arrow(
    location = "bl",
    pad_x = unit(0, "in"),
    pad_y = unit(0.3, "in"),
    style = north_arrow_fancy_orienteering)+
    theme_bw()

p1


plot_grid(p2, p1, labels = 'AUTO',ncol = 2,rel_widths =c(1,5))


#plot(coord.pop, xlab = "Longitude", ylab = "Latitude", type = "n")
#map(add = T, col = "grey90", fill = TRUE)

#for (i in 1:length(pops)){
#  add.pie(z = qpop[i,], x = coord.pop[i,1], y = coord.pop[i,2], labels = "",
#  col = c("orange","violet","lightgreen","yellow","purple","lightblue"))}
```


DAPC
```{r}
pca <- glPca(neutral_genlight)

#choosing k=9
dapc <- dapc(neutral_genlight, n.pca = 3, n.da = 2)

scatter(dapc, col = group.colors, cex = 2, legend = TRUE, clabel = F, posi.leg = "bottomleft", scree.pca = TRUE,
        posi.pca = "topleft", cleg = 0.75)

#compoplot(dapc)

dapc.results <- as.data.frame(dapc$posterior)
dapc.results$pop <- pop(neutral_genlight)
dapc.results$indNames <- rownames(dapc.results)

dapc.results <- pivot_longer(dapc.results, -c(pop, indNames))

colnames(dapc.results) <- c("Original_Pop","Sample","Assigned_Pop","Posterior_membership_probability")

p <- ggplot(dapc.results, aes(x=Sample, y=Posterior_membership_probability, fill=Assigned_Pop))+ 
  geom_bar(stat='identity') +
  facet_grid(~Original_Pop, scales = "free") +
  theme(axis.text.x=element_blank(),axis.ticks.x=element_blank())+
  scale_fill_manual(values=group.colors)
p
```

a tree of pops 
```{r}
tree <- aboot(neutral_genpop,tree = "upgma", sample = 100,cutoff = 50)

#cols <- hcl.colors(18, "Zissou1", rev = FALSE)

plot.phylo(tree, cex = 0.8, font = 2, adj = 0, tip.color =  group.colors)
nodelabels(tree$node.label, adj = c(1.3, -0.5), frame = "n", cex = 0.8,font = 3, xpd = TRUE)
axis(side = 1)
title(xlab = "Genetic distance (proportion of loci that are different)")
```

triangle plot test. probably not the move for this project
```{r}
library(dartR)
library(vcfR)
#dartR::gl.install.vanilla.dartR(flavour="dev")

tri_test <- neutral_matrix[grep("NESR", rownames(neutral_matrix)), , drop = FALSE]

tri_test_genind <- df2genind(tri_test, sep ="", ploidy=2,NA.char = "00")

tri_test_genlight <- gi2gl(tri_test_genind, parallel = FALSE, verbose = NULL)
ploidy(tri_test_genlight) <- 2

#DB to try to add sex and chr to this eventually?
#gota conver to vcf
setwd("C:/Users/blumsted/Documents")
tri_test_vcf<-gl2vcf(tri_test_genlight,outpath = getwd(),plink_path = getwd())

setwd("~/projects/ots_coastal")
library(vcfR)
#devtools::install_github("omys-omics/triangulaR")
library(triangulaR)

data <- read.vcfR("gl_vcf.vcf", verbose = F)
data

tri_test_meta <- dplyr::filter(meta, grepl('NESR', sample))

popmap <- data.frame(
  id = paste("pop1_",tri_test_meta$sample,sep=""),
  pop = tri_test_meta$run
)

popmap[is.na(popmap)] <- "unk"

vcfR.diff <- alleleFreqDiff(vcfR = data, pm = popmap, p1 = "spring", p2 = "fall", difference = 0.9)

hi.het <- hybridIndex(vcfR = data, pm = popmap, p1 = "spring", p2 = "fall")

triangle.plot(hi.het)
missing.plot(hi.het)
```

